sparse/coo/umath_cython.pyx

```diff
@@ -0,0 +1,57 @@
+import numpy as np
+cimport numpy as np
+cimport cython
+from libcpp.vector cimport vector
+from numpy cimport int64_t
+
+cdef extern from "<cstring>" namespace "std":
+    void* memcpy (void* destination, const void* source, size_t num)
+
+
+def cywhere(np.ndarray a not None, np.ndarray b not None):
+    a = a.astype(dtype=np.int64)
+    b = b.astype(dtype=np.int64)
+
+    a_ind, b_ind = cywhere_wrapped(a, b)
+
+    a_ind = a_ind.astype(np.min_scalar_type(a.shape[0]))
+    b_ind = b_ind.astype(np.min_scalar_type(b.shape[0]))
+
+    return a_ind, b_ind
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef cywhere_wrapped(int64_t[:] a, int64_t[:] b):
+    cdef vector[int64_t] a_ind = vector[int64_t]()
+    cdef vector[int64_t] b_ind = vector[int64_t]()
```

Comment:
If you know these sizes ahead of time I recommend pre-allocating these as static numpy arrays (or something similar).

Reply:
Or even if you're only able to bound the size reasonably.

Reply:
Unfortunately, I don't. It could be zero, it could be `na * nb` in the worst case. And having a "size" pass that just counts the size, then a second pass, makes it go from `6.55` to `~8` seconds.

Reply:
Is there a particular benchmark that you're using?

Reply:
Here it is:

```python
In[2]: import numpy as np
In[3]: import sparse
In[4]: x = sparse.random((1000, 1000, 1000), density=0.01)
In[5]: y = sparse.random((1000, 1000, 1000), density=0.01)
In[6]: %timeit x + y
```

Reply:
Just so you know, this function takes in two sorted 1-D `ndarray`s, and performs the equivalent of an SQL outer join on them.

Reply:
That's useful context, thanks!

I tried rewriting this in Numba using two passes and got similar results. https://gist.github.com/35f0d359f98b370c5a7b2cd862333577

Throwing in a couple print statements:

```python
In [1]: import sparse

In [2]: x = sparse.random((1000, 1000, 1000), density=0.01)

In [3]: y = sparse.random((1000, 1000, 1000), density=0.01)

In [4]: x + y
duration: 1.1437668800354004
n 100564 m 993185491
duration: 1.1581335067749023
duration: 1.039297342300415
n 100564 m 993185491
duration: 1.030043125152588
duration: 1.0377395153045654
n 100564 m 993774438
duration: 1.040597677230835
Out[4]: <COO: shape=(1000, 1000, 1000), dtype=float64, nnz=19899436, sorted=False, duplicates=False>
```

It's unfortunate that we seem to be passing through all billion possibilities.  Each pass takes around 1s, which naively matches what I would expect from a sequential operation on a GHz processor.  The fact that there are three calls with the same length is interesting, implying that we might be able to drop the first pass in the second two cases (although this still may not compete against the dynamic approach).  Perhaps there is some indexing structure that we can save in order to avoid the second pass as well?  I guess I'm curious why we call this function three times, and if there is work that can be saved between those subsequent calls (or if they're even necessary).

I also haven't looked at all at the algorithm to see if it's ideal.

Reply:
> It's unfortunate that we seem to be passing through all billion possibilities.

We aren't. [See my SO question on the topic](https://stackoverflow.com/questions/47831287/get-index-of-matching-elements-in-array-taking-into-account-repetitions) to see what the algorithm does. TL;DR, it moves forward in each array of nonzeros, only adding those indices into the original arrays that match. It's linear in the output size, so it's as efficient as can be made.

For `operator.add`: One match for "common" elements, and two of the others to filter out "already matched" so the "overlap" isn't operated on with zero. As far as I can tell, this was necessary for broadcasting.

Reply:
In my case the nnz was around 2e7 while the number of passes through the doubly nested loop was 1e9.  So the code that I wrote (which I think is similar to the code that you wrote) is not linear in the output size.